README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ransid <img src="man/figures/logo.png" align="right" height=450/>

<!-- badges: start -->

![](https://img.shields.io/badge/cool-useless-green.svg) [![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

`ransid` converts images to ANSI text which can be displayed in the
console.

  - Supports 8-bit and 24-bit ANSI (but Rstudio terminal only supports
    8-bit colour)
  - Specify colours using standard R colour names e.g. ‘hotpink’,
    ‘skyblue’ and `ransid` will map them to the nearest available ANSI
    colour

## Installation

You can install `ransid` from
[github](https://github.com/coolbutuseless/ransid) with:

``` r
# install.packages('remotes')
remotes::install_github("coolbutuseless/ransid")
```

## Example - 216 colour ANSI

**Note** - because ANSI is for display in a terminal, it does not render
in a markdown document. The following images are screenshots of my
Rstudio terminal

``` r
im <- magick::image_read(system.file('img', 'Rlogo.png', package = 'png'))
cat(im2ansi(im, width = 120))
```

<img src="man/figures/R.png" width="100%" />

``` r
im <- magick::image_read('https://www.fillmurray.com/300/250')
cat(im2ansi(im, width = 120))
```

<img src="man/figures/murray.png" width="100%" />

``` r
im <- magick::image_read('https://placekitten.com/300/200')
cat(im2ansi(im, width = 120))
```

<img src="man/figures/kitten.png" width="100%" />

## Example - 24bit ANSI colour

24bit colour ANSI is enabled by setting `full_colour = TRUE`

24bit ANSI colour has more limited support (compared to 8bit ANSI), and
will **not** display in Rstudio.

It will however work in `iTerm` on OSX, KDE Konsole, Xterm, libvte-based
terminals such as GNOME Terminal. See [the wikipedia
page](https://en.wikipedia.org/wiki/ANSI_escape_code) for more
information

``` r
im <- magick::image_read('https://placekitten.com/300/200')
cat(im2ansi(im, width = 200, full_colour = TRUE))
```

<img src="man/figures/kitten24.png" width="100%" />

## Access raw ANSI

Lower level functions are also available.

  - `col2bg`, `col2bg24` - create an ANSI sequence for a particular
    background colour (8bit and 24bit versions)
  - `col2fg`, `col2fg24` - create an ANSI sequence for a particular
    foreground colour (8bit and 24bit versions)
  - `im2char` - the matrix corresponding to the image, where each
    element represents one pixel

#### `col2bg`

``` r
col2bg('tomato')
#> [1] "\033[48;5;209m"
paste0("Hello ", col2bg("tomato"), "#RStats", reset_code)
#> [1] "Hello \033[48;5;209m#RStats\033[39m\033[49m"
```

``` r
cat(paste0("Hello ", col2bg("tomato"), "#RStats", reset_code), "\n")
```

<img src="man/figures/text1.png" width="20%" />

#### `col2fg`

``` r
col2fg('darkgreen')
#> [1] "\033[38;5;28m"
paste0("Hello ", col2bg('tomato'), col2fg('darkgreen'), "#RStats", reset_code)
#> [1] "Hello \033[48;5;209m\033[38;5;28m#RStats\033[39m\033[49m"
```

``` r
cat(paste0("Hello ", col2bg('tomato'), col2fg('darkgreen'), "#RStats", reset_code), "\n")
```

<img src="man/figures/text2.png" width="20%" />

#### `im2char`

``` r
im <- image_read(system.file('img', 'Rlogo.png', package = 'png'))
char_matrix <- im2char(im)
char_matrix[1, 1:10]
#>  [1] "\033[48;5;231m " "\033[48;5;231m " "\033[48;5;231m " "\033[48;5;231m "
#>  [5] "\033[48;5;231m " "\033[48;5;231m " "\033[48;5;231m " "\033[48;5;231m "
#>  [9] "\033[48;5;231m " "\033[48;5;231m "
```

#### `col2bg24`

``` r
col2bg24('tomato')
#> [1] "\033[48;2;255;99;71m"
paste0("Hello ", col2bg24("tomato"), "#RStats", reset_code)
#> [1] "Hello \033[48;2;255;99;71m#RStats\033[39m\033[49m"
```

``` r
cat(paste0("Hello ", col2bg24('tomato'), col2fg24('darkgreen'), "#RStats", reset_code), "\n")
```

<img src="man/figures/text3.png" width="20%" />

# Related Software

Some great R packages for ANSI handling that eclipse what this humble
package does:

  - [crayon](https://cran.r-project.org/package=crayon) the great,
    useful and well-designed ANSI package by Gábor Csárdi and Brodie
    Gaslam
  - [fansi](https://cran.r-project.org/web/packages/fansi/index.html)
    Brodie Gaslam’s comprehensive set of utilities for wrangling strings
    containing ANSI escape sequences.


NEWS.md
# ransid 0.1.3 2020-10-16

* `NA` colours return an emptry string rather than white

# ransid 0.1.2

* Include 24bit ANSI output with `full_colour` argument
* Vectorised the 8bit ANSI escape code lookup

# ransid 0.1.1

* Tweak interface

# ransid 0.1.0

* Initial release


R/image.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert an image to an ANSI string
#'
#' @param im magick image
#' @param width text output width. default: 80
#' @param font_aspect ratio of width to height for the output font. default: 0.45
#' @param full_colour use 24bit colour codes. default: FALSE. Only supported by
#'        some terminals
#'
#' @return Character string with ANSI colours
#'
#' @import magick
#' @importFrom grDevices as.raster
#' @export
#'
#' @examples
#' \dontrun{
#' im <- image_read(system.file('img', 'Rlogo.png', package = 'png'))
#' cat(im2ansi(im, width = 80))
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
im2ansi <- function(im, width = 80, font_aspect = 0.45, full_colour = FALSE) {

  mat <- im2char(im, width, font_aspect, full_colour)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Collapse
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  rows <- apply(mat, 1, paste0, collapse = "")
  rows <- paste0(rows, reset_code, "\n")
  paste0(rows, collapse = "")
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert an image to a matrix of ANSI chars
#'
#' @inheritParams im2ansi
#'
#' @return Character string with ANSI colours
#'
#' @import magick
#' @importFrom grDevices as.raster
#' @export
#'
#' @examples
#' \dontrun{
#' im <- image_read(system.file('img', 'Rlogo.png', package = 'png'))
#' mat <- im2char(im)
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
im2char <- function(im, width = 80, font_aspect = 0.45, full_colour = FALSE) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Scale the image:
  #    * for the given font_aspect
  #    * for the given output width
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  info         <- image_info(im)
  width_factor <- width/info$width
  final_height <- info$height * width_factor * font_aspect
  final_height <- as.integer(round(final_height))

  im <- magick::image_resize(im, geometry = paste0(width, "x", final_height, "!"))

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Convert to raster
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ras <- grDevices::as.raster(im)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Get the ANSI match for each colour, and use it to colour a plain 'space'
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (full_colour) {
    ansi <- col2bg24(ras)
  } else {
    ansi <- col2bg(ras)
  }
  ansi <- paste0(ansi, ' ')

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Reshape for return to the user
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  dim(ansi) <- rev(dim(ras))
  ansi <- t(ansi)

  ansi
}





if (FALSE) {
  im <- image_read(system.file('img', 'Rlogo.png', package = 'png'))
  res <- im2ansi(im, width = 80, full_colour = TRUE)
  cat(res)

  ff <- system.file('img', 'magpie.jpg', package = 'ggpattern')
  im <- image_read(ff)
  cat(im2ansi(im))
}


R/ansi-escape-codes-24.R

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert an R colour to an ansi string for 24bit colour supported by some terminals
#'
#' Ref: \url{https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit}
#'
#' @param rcolour any R colour e.g. 'red', '#445566'
#'
#' @return ANSI escape string for the given colour as a foreground or background
#'         colour
#'
#' @importFrom grDevices col2rgb
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
col2bg24 <- function(rcolour) {
  ifelse(
    is.na(rcolour),
    '',
    {
      cols <- col2rgb(rcolour)
      paste0("\033[48;2;", cols[1,], ";", cols[2,], ";", cols[3,], "m")
    }
  )
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname col2bg24
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
col2fg24 <- function(rcolour) {
  ifelse(
    is.na(rcolour),
    '',
    {
      cols <- col2rgb(rcolour)
      paste0("\033[38;2;", cols[1,], ";", cols[2,], ";", cols[3,], "m")
    }
  )
}


R/ansi-escape-codes.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Reset back to terminal defaults
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
reset_code <- "\033[39m\033[49m"

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert an R colour to a 216-colour ANSI string. Suitable for most terminals
#'
#' Ref: \url{https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit}
#'
#' If all the RGB colour components are equal then the colour is matched to one
#' of 24 grey levels, other wise it is converted to one of 216 standard colours.
#'
#' @param rcolour any R colour e.g. 'red', '#445566'
#'
#' @return ANSI escape string for the given colour as a foreground or background
#'         colour
#'
#' @importFrom grDevices col2rgb
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
col2bg <- function(rcolour) {
  ifelse(is.na(rcolour), '', paste0("\033[48;5;", col2code(rcolour), "m"))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname col2bg
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
col2fg <- function(rcolour) {
  ifelse(is.na(rcolour), '', paste0("\033[38;5;", col2code(rcolour), "m"))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname col2bg
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
col2code <- function(rcolour) {
  cols <- col2rgb(rcolour)

  is_grey        <- cols[1,] == cols[2,] & cols[2,] == cols[3,]
  possibly_white <- cols[1,] == 255L

  grey_code <- 232L + as.integer(round(cols[1,]/255 * 23))

  cols <- round(cols/255 * 5)
  colour_code <- 16L + 36L * cols[1,] + 6 * cols[2,] + cols[3,]

  ifelse(is_grey & !possibly_white, grey_code, colour_code)
}

